airbyte-integrations/connectors/source-google-sheets/source_google_sheets/source.py

```diff
@@ -165,46 +169,61 @@ def _read(
+                yield as_airbyte_message(stream, AirbyteStreamStatus.COMPLETE)
+                continue
+
+            column_index_to_name = sheet_to_column_index_to_name[sheet]
+            row_cursor = 2  # we start syncing past the header row
+            # For the loop, it is necessary that the initial row exists when we send a request to the API,
+            # if the last row of the interval goes outside the sheet - this is normal, we will return
+            # only the real data of the sheet and in the next iteration we will loop out.
+            while row_cursor <= sheet_row_counts[sheet]:
+                row_batch = SpreadsheetValues.parse_obj(
+                    client.get_values(
+                        sheet=sheet,
+                        row_cursor=row_cursor,
+                        spreadsheetId=spreadsheet_id,
+                        majorDimension="ROWS",
                     )
+                )
 
-                    row_cursor += client.Backoff.row_batch_size + 1
-                    # there should always be one range since we requested only one
-                    value_ranges = row_batch.valueRanges[0]
+                row_cursor += client.Backoff.row_batch_size + 1
+                # there should always be one range since we requested only one
+                value_ranges = row_batch.valueRanges[0]
 
-                    if not value_ranges.values:
-                        break
+                if not value_ranges.values:
+                    break
 
-                    row_values = value_ranges.values
-                    if len(row_values) == 0:
-                        break
+                row_values = value_ranges.values
+                if len(row_values) == 0:
+                    break
 
-                    yield as_airbyte_message(stream, AirbyteStreamStatus.RUNNING)
-                    for row in row_values:
-                        if not Helpers.is_row_empty(row) and Helpers.row_contains_relevant_data(row, column_index_to_name.keys()):
-                            yield AirbyteMessage(
-                                type=Type.RECORD, record=Helpers.row_data_to_record_message(sheet, row, column_index_to_name)
-                            )
-                yield as_airbyte_message(stream, AirbyteStreamStatus.COMPLETE)
-            else:
-                logger.info(f"Skipping syncing sheet {sheet}: {reason}")
+                yield as_airbyte_message(stream, AirbyteStreamStatus.RUNNING)
+                for row in row_values:
+                    if not Helpers.is_row_empty(row) and Helpers.row_contains_relevant_data(row, column_index_to_name.keys()):
+                        yield AirbyteMessage(type=Type.RECORD, record=Helpers.row_data_to_record_message(sheet, row, column_index_to_name))
+
+            yield self._checkpoint_state(checkpoint_reader.get_checkpoint(), state_manager, sheet, None)
+            yield as_airbyte_message(stream, AirbyteStreamStatus.COMPLETE)
```

Comment:
Please add error handling for per-stream error, like it is implemented in class AbstractSource

Reply:
does it affect emitting a state message? I'd rather implement it in a separate ticket if it does not